File: legacy-nodejs/src/repositories/user.repository.ts
import { PrismaClient } from '@prisma/client';
import { prisma } from '../config/database';
import { IUserRepository } from './interfaces';
import { UserRecord, Role } from '../types/domain';

// The ONLY layer permitted to issue Prisma queries for the User entity
export class UserRepository implements IUserRepository {
  constructor(private readonly client: PrismaClient = prisma) {}

  async findByEmail(email: string): Promise<UserRecord | null> {
    const user = await this.client.user.findUnique({ where: { email } });
    return user ? this.toDomain(user) : null;
  }

  async findById(id: string): Promise<UserRecord | null> {
    const user = await this.client.user.findUnique({ where: { id } });
    return user ? this.toDomain(user) : null;
  }

  async create(input: { email: string; passwordHash: string; fullName: string }): Promise<UserRecord> {
    const user = await this.client.user.create({
      data: {
        email: input.email,
        passwordHash: input.passwordHash,
        fullName: input.fullName
        // role intentionally omitted and it defaults to 'customer' at the schema
      }
    });
    return this.toDomain(user);
  }

  async updateProfile(id: string, data: { fullName?: string }): Promise<UserRecord> {
    const user = await this.client.user.update({ where: { id }, data });
    return this.toDomain(user);
  }

  private toDomain(row: any): UserRecord {
    return {
      id: row.id,
      email: row.email,
      passwordHash: row.passwordHash,
      fullName: row.fullName,
      role: row.role as Role,
      createdAt: row.createdAt
    };
  }
}

File: src/Repositories/Implementations/UserRepository.cs
using ECommerceApi.Data;
using ECommerceApi.Models;
using ECommerceApi.Repositories.Interfaces;
using Microsoft.EntityFrameworkCore;

namespace ECommerceApi.Repositories.Implementations;

// The ONLY layer permitted to query AppDbContext for User persistence
// Contains no business rules
public class UserRepository : IUserRepository
{
    private readonly AppDbContext _context;

    public UserRepository(AppDbContext context)
    {
        _context = context;
    }

    public async Task<User?> FindByEmailAsync(string email)
    {
        return await _context.Users.AsNoTracking().FirstOrDefaultAsync(u => u.Email == email);
    }

    public async Task<User?> FindByIdAsync(Guid id)
    {
        return await _context.Users.AsNoTracking().FirstOrDefaultAsync(u => u.Id == id);
    }

    public async Task<User> CreateAsync(string email, string passwordHash, string fullName)
    {
        var user = new User
        {
            Email = email,
            PasswordHash = passwordHash,
            FullName = fullName
            // Role intentionally omitted. Defaults to Customer at the  model level. Nothing client-supplied can set role at registration time 
        };

        _context.Users.Add(user);
        await _context.SaveChangesAsync();
        return user;
    }

    public async Task<User> UpdateProfileAsync(Guid id, string? fullName)
    {
        var user = await _context.Users.FirstAsync(u => u.Id == id);
        if (fullName is not null)
        {
            user.FullName = fullName;
        }
        await _context.SaveChangesAsync();
        return user;
    }
}
